lib/i18n_flat_yml_backend.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+require 'i18n'
+require 'json'
+
+# Custom i18n backend that parse our "flat_yml" files into the nested
+# hash structure that i18n works with
+class I18nFlatYmlBackend < I18n::Backend::Simple
+  # @param filename [String] filename, assumed to have the locale slug in the filename ex "en.txt"
+  # @return [Array(Hash, Boolean)] tuple of a hash and keys_symbolized
+  def load_yml(filename)
+    content, keys_symbolized = super
+
+    if self.class.nested_hashes?(content)
+      [content, keys_symbolized]
+    else
+      [
+        {
+          self.class.locale(filename) => self.class.unflatten(content),
+        },
+        false,
+      ]
+    end
+  end
+
+  # @example
+  #   locale("config/locales/fr.yml")
+  #   # => "fr"
+  def self.locale(filename)
+    File.basename(filename, '.yml')
+  end
+
+  # @return [Boolean] true if +content+ appears to be a legacy "nested" yml file
+  #   instead of a flat yml file
+  def self.nested_hashes?(content)
+    content.is_a?(Hash) && content.keys.size == 1 && content[content.keys.first].is_a?(Hash)
+  end
+
+  # @param [Hash<String, String>] key_values
+  # @return [Hash<String, Hash<String, String>>]
+  def self.unflatten(key_values)
+    result = {}
+
+    key_values.each do |full_key, value|
+      *key_parts, last = full_key.to_s.split('.')
+
+      to_insert = result
+
+      key_parts.each do |key_part|
+        to_insert = (to_insert[key_part] ||= {})
+      end
+
+      to_insert[last] = value
+    end
+
+    convert_arrays(result)
+  end
+
+  # If all keys of a hash are numeric, converts the hash to an array
+  def self.convert_arrays(outer)
+    outer.transform_values do |inner|
+      if inner.is_a?(Hash)
+        if inner.keys.all? { |key| numeric_key?(key) }
```

Comment:
Would it be possible / reasonable to let the arrays exist as arrays within the YAML? I'm also wondering if we could get to a point where we don't need to unflatten this file and could instead use it as a flattened hash, where array usage in code seems like one of the few issues to need to deal with.

Reply:
I don't feel that strongly either way? I thought having arrays was nice. Is your worry that the code is kind of clunky?

Reply:
I think my feelings here are:

- The convention of `.0`, `.1`, etc. suffixes being transformed to an array automagically may not be obvious to a new developer
- Arrays could create an easier migration path to avoiding the unflattening, where expected usage of array as `t('foo.bar').each` maps directly to the key `foo.bar` in the flattened set of keys (instead of `foo.bar.0`, etc.)
- Limit the amount of transformative work we have to do in the backend code

Reply:
I guess luckily we use not that many arrays in these translation files so there's less risk of these getting caught in merge conflicts, I'll remove this special-cased array handling and use YAML array syntax